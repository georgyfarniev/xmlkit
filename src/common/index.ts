export * from './types'
export * from './stack'
export * from './serializer'
