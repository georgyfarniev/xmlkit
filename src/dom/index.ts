export * from './xml_cdata'
export * from './xml_comment'
export * from './xml_document'
export * from './xml_element'
export * from './xml_node'
