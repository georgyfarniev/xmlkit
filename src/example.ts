import { XmlDocument } from './xml_document'

const XML_TEXT = `
<?xml version="1.0" encoding="UTF-8"?>
<Root>
  <Child1 foo="bar">Text</Child1>
  <Child2>Text</Child2>
  <Child2>
    <!-- Comment2! -->
    <Grandchild1 fiz="buz" foo="123">Text3</Grandchild1>
    <Grandchild2>Text4</Grandchild2>
  </Child2>
  <Child3/>
  <Child4>
    <LastChild>Text</LastChild>
  </Child4>
  <!-- Comment! -->
</Root>`

async function main() {
  const doc = new XmlDocument(XML_TEXT)

  for (const child of doc.root) {
    console.log(child.name)

    for (const subchild of child) {
      console.log(`\t${subchild.name}`)
    }
  }
}

if (require.main === module) {
  main()
}
