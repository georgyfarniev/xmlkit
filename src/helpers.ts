import { Readable } from 'stream'

/**
 * Class to generate heavy xml payload for testing purposes.
 * Each value inside contains incremented value for ease of testing
 */
export class TestStream extends Readable {
  private index = 0

  constructor(private readonly total: number = 1 * 1000 * 1000) {
    super()
  }

  public _read(): void {
    const { index, total } = this

    if (index > total) {
      this.push(null)
      return
    }

    const buf = [
      index ? '' : '<root>',
      `<item foo="bar${index}">`,
      `<node1_${index}>text${index}`,
      `<subnode${index}>subtext${index}</subnode${index}>`,
      `</node1_${index}>`,
      '</item>',
      index === total ? '</root>' : ''
    ].join('\n')

    this.push(buf)
    this.index++
  }
}
