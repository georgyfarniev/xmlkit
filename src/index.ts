
export * from './dom'
export * from './xml_sax'
export * from './xml_stream'
