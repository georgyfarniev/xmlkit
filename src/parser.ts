import sax from 'sax'
import type { Tag } from 'sax';
import {
  IXmlCData,
  IXmlCloseTag,
  IXmlComment,
  IXmlOpenTag,
  IXmlText,
  XmlToken,
  XmlTokenType,
  BaseParser
} from './types';


// Abstraction class to hide sax module behind and allow easy replacement
export class XmlSaxParser implements BaseParser {
  private tokens: XmlToken[] = [];

  constructor(private parser = sax.parser(true)) {
    parser.onopentag = this.beginElement
    parser.onclosetag = this.endElement
    parser.ontext = this.onText
    parser.oncomment = this.onComment
    parser.oncdata = this.onCDATA
  }
 
  //#region  handlers
  private beginElement = (tag: Tag) => {
    const { name, attributes, isSelfClosing } = tag;
    const tok: IXmlOpenTag = {
      name, 
      attrs: attributes,
      selfClosing: isSelfClosing,
      type: XmlTokenType.ElementOpen
    };

    this.tokens.push(tok);
  }

  private endElement = (name: string) => {
    this.tokens.push({
      name,
      type: XmlTokenType.ElementClose
    } as IXmlCloseTag);
  }

  private onText = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.Text
    } as IXmlText);
  }

  private onCDATA = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.CDATA
    } as IXmlCData);
  }

  private onComment = (text: string) => {
    this.tokens.push({
      text,
      type: XmlTokenType.Comment
    } as IXmlComment);
  }
  //#endregion

  public getTokens(chunk: string): XmlToken[] {
    this.tokens = []
    this.parser.write(chunk)
    return this.tokens;
  }
}
