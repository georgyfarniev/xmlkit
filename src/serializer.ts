import { Attrs } from './types';

export class XmlSerializer {
  public static openTag(
    name: string,
    attrs: Attrs = {},
    selfClosing = false
  ): string {
    const aStr = Object.entries(attrs || [])
      .reduce((acc, [k, v]) => `${acc}${k}="${v}" `, ' ')
      .trimRight();

    const sc = selfClosing ? '/' : '';
    return `<${name}${aStr}${sc}>`;
  }

  public static closeTag(name: string): string {
    return `</${name}>`;
  }

  public static text(text: string): string {
    return text;
  }

  public static comment(text: string): string {
    return `<!-- ${text} -->`
  }

  public static CDATA(text: string): string {
    return `<![CDATA[${text}]]>`
  }
}