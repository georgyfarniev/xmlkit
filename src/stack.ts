/**
 * Stack implementation
 */

export class Stack<T = unknown> {
  private readonly data: T[] = []

  public toArray(): T[] { return this.data }
  public push(elt: T): void { this.data.push(elt) }
  public pop(): T { return this.data.pop() }
  public clear(): void { this.data.length = 0 }
  public get length(): number { return this.data.length }
  public get empty(): boolean { return this.data.length === 0 }

  public get top() :T|undefined {
    return this.data.length ? this.data[this.data.length - 1] : undefined
  }
}
