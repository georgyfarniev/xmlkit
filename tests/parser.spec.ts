import 'jest';
import { XmlSaxParser } from '../src/parser';

describe('XmlSaxParser', () => {
  test('simple', () => {
    const parser = new XmlSaxParser();
    const xml = '<root><element>text</element></root>'
    expect(parser.getTokens(xml)).toMatchSnapshot();
  });

  test('complex', () => {
    const parser = new XmlSaxParser();
    const xml = '<root><element foo="bar">text<inner fiz="buzz"/></element></root>'
    expect(parser.getTokens(xml)).toMatchSnapshot();
  });
});
