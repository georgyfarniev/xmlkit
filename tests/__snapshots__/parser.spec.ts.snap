// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XmlSaxParser complex 1`] = `
[
  {
    "attrs": {},
    "name": "root",
    "selfClosing": false,
    "type": "element_open",
  },
  {
    "attrs": {
      "foo": "bar",
    },
    "name": "element",
    "selfClosing": false,
    "type": "element_open",
  },
  {
    "text": "text",
    "type": "text",
  },
  {
    "attrs": {
      "fiz": "buzz",
    },
    "name": "inner",
    "selfClosing": true,
    "type": "element_open",
  },
  {
    "name": "inner",
    "type": "element_close",
  },
  {
    "name": "element",
    "type": "element_close",
  },
  {
    "name": "root",
    "type": "element_close",
  },
]
`;

exports[`XmlSaxParser simple 1`] = `
[
  {
    "attrs": {},
    "name": "root",
    "selfClosing": false,
    "type": "element_open",
  },
  {
    "attrs": {},
    "name": "element",
    "selfClosing": false,
    "type": "element_open",
  },
  {
    "text": "text",
    "type": "text",
  },
  {
    "name": "element",
    "type": "element_close",
  },
  {
    "name": "root",
    "type": "element_close",
  },
]
`;
